backend/x86_binary_emitter.ml

```diff
@@ -1205,6 +1336,26 @@ let emit_cqto b =
   emit_rex b rexw;
   buf_int8 b 0x99
 
+let emit_crc32 b ~dst ~src =
+  match (dst, src) with
+  | (Reg32 reg, ((Reg8L _ | Reg8H _ | Mem {typ = BYTE; _} | Mem64_RIP (BYTE, _, _)) as rm)) ->
+    (* CRC32 r32, r/m8 *)
+    buf_int8 b 0xF2;
+    emit_mod_rm_reg b rex [ 0x0F; 0x38; 0xF0 ] rm (rd_of_reg64 reg)
+  | (Reg64 reg, ((Reg8L _ | Reg8H _ | Mem {typ = BYTE; _} | Mem64_RIP (BYTE, _, _)) as rm)) ->
+    (* CRC32 r64, r/m8 *)
+    buf_int8 b 0xF2;
+    emit_mod_rm_reg b rexw [ 0x0F; 0x38; 0xF0 ] rm (rd_of_reg64 reg)
+  | (Reg32 reg, ((Reg16 _ | Reg32 _ | Mem _ | Mem64_RIP _) as rm)) ->
```

Comment:
does the `typ` field of `Mem` need to be constraint here to `DWORD` or `QWORD`?

Reply:
That's a very good question. I was not sure what to do in those cases so I followed those rules, which seems to be consistent with what the emitter was already doing:
1. When it's non ambiguous, for instance because the dest operand is different in all pattern matching cases, I don't restrict `Mem _ | Mem64_RIP _`'s `data_type`, assuming the compiler won't produce nonsensical values
2. When it needs disambiguation, I match over mem operands of the right `data_type`.

My main problem is that I'm not sure exactly how `data_type` is used by the compiler since in the parts of the reference manual I had to deal with they only ever mentioned the size (byte, 16, 32, 64 or 128-bits) of the memory operands and I am still a bit confused by the other variants in `data_type`, i.e. `NONE | REAL4 | REAL8 | NEAR | PROC`.

If you have any input on this that would be extremely useful.

Reply:
Ok, this case doesn't need disambiguation based on what the compiler can generate now.

`REAL4` and `REAL8` are single and double precision floats (so 32 and 64 bits, resp). I don't exactly what the others mean. A quick search reveals that NEAR is only used in x86_masm and needed for labels, PROC is only used in masm for extern for i386, NONE is used for LEA instruction and labels.